evmcore/validation_test.go

```diff
@@ -0,0 +1,620 @@
+		opt := testTransactionsOption()
+
+		// setup tx to fail intrinsic gas calculation
+		setData(t, tx, data[:txSlotSize])
+		floorDataGas, err := core.FloorDataGas(types.NewTx(tx).Data())
+		require.NoError(t, err)
+		setGas(t, tx, floorDataGas-1)
+		opt.currentMaxGas = floorDataGas
+
+		// --- needed for execution up to relevant check ---
+		// set tx for execution
+		setGasFeeCap(t, tx, opt.currentGasPrice)
+		// sign txs with sender
+		signer, address, signedTx := testSignTx(t, tx)
+		// setup enough balance
+		testDb := newTestTxPoolStateDb()
+		testDb.balances[address] = uint256.NewInt(math.MaxUint64)
+		opt.currentState = testDb
+		// ---
+
+		// validate transaction
+		err = validateTx(signedTx, signer, opt)
+		require.ErrorIs(t, err, ErrFloorDataGas)
+	})
+}
+
+func TestValidation_validateTx_Success(t *testing.T) {
+	tests := []types.TxData{
+		&types.LegacyTx{
+			Nonce:    0,
+			GasPrice: big.NewInt(1),
+			Gas:      21000,
+			To:       &common.Address{},
+			Value:    big.NewInt(1),
+		},
+		&types.AccessListTx{
+			Nonce:      0,
+			GasPrice:   big.NewInt(1),
+			Gas:        21000,
+			To:         &common.Address{},
+			Value:      big.NewInt(1),
+			AccessList: types.AccessList{},
+		},
+		&types.DynamicFeeTx{
+			Nonce:     0,
+			GasTipCap: big.NewInt(1),
+			GasFeeCap: big.NewInt(2),
+			Gas:       21000,
+			To:        &common.Address{},
+			Value:     big.NewInt(1),
+		},
+		&types.BlobTx{
+			Nonce:     0,
+			GasTipCap: uint256.NewInt(1),
+			GasFeeCap: uint256.NewInt(2),
+			Gas:       21000,
+		},
+		&types.SetCodeTx{
+			Nonce:     0,
+			GasTipCap: uint256.NewInt(1),
+			GasFeeCap: uint256.NewInt(2),
+			Gas:       46000, // needs more gas than other tx types because of the auth list
+			AuthList:  []types.SetCodeAuthorization{{}},
+		},
+	}
+
+	for _, tx := range tests {
+		t.Run(txTypeName(tx), func(t *testing.T) {
+			// Sign the transaction
+			signer, address, signedTx := testSignTx(t, tx)
+
+			// Set up sufficient balance and nonce
+			testDb := newTestTxPoolStateDb()
+			testDb.balances[address] = uint256.NewInt(math.MaxUint64)
+			testDb.nonces[address] = 0
+
+			opts := testTransactionsOption()
+			opts.currentState = testDb
+
+			// Validate the transaction
+			err := validateTx(signedTx, signer, opts)
+			require.NoError(t, err)
+		})
+	}
+}
+
+////////////////////////////////////////////////////////////////////////////////
+// Helper functions for testing.
+
+// allTxTypes returns a list of all transaction types for testing.
+func allTxTypes() []types.TxData {
+	return []types.TxData{
+		&types.LegacyTx{},
+		&types.AccessListTx{},
+		&types.DynamicFeeTx{},
+		makeBlobTx(nil, nil),
+		&types.SetCodeTx{AuthList: []types.SetCodeAuthorization{{}}},
+	}
+}
+
+// testSignTx generates a new key, signs the transaction with it, and returns
+// the signer, address, and signed transaction.
+func testSignTx(t *testing.T, tx types.TxData) (types.Signer, common.Address, *types.Transaction) {
```

Comment:
```suggestion
func signTx(t *testing.T, tx types.TxData) (types.Signer, common.Address, *types.Transaction) {
```

The reason is that one may think that this function is testing signing

Reply:
what do you think about `signTxForTest` ? considering there is already a function called `signTx` and it is used already in this package, I would like to keep the names as different/explicit as possible to prevent confusion.